--- tools/hook-scripts/commit-email.pl	2004-06-14 22:29:22.000000000 +0200
+++ tools/hook-scripts/commit-email.pl	2005-10-30 22:39:22.942246752 +0100
@@ -222,32 +222,16 @@
 shift @svnlooklines;
 my @log = map { "$_\n" } @svnlooklines;
 
-# Figure out what directories have changed using svnlook.
-my @dirschanged = &read_from_process($svnlook, 'dirs-changed', $repos, 
-                                     '-r', $rev);
-
-# Lose the trailing slash in the directory names if one exists, except
-# in the case of '/'.
-my $rootchanged = 0;
-for (my $i=0; $i<@dirschanged; ++$i)
-  {
-    if ($dirschanged[$i] eq '/')
-      {
-        $rootchanged = 1;
-      }
-    else
-      {
-        $dirschanged[$i] =~ s#^(.+)[/\\]$#$1#;
-      }
-  }
 
 # Figure out what files have changed using svnlook.
 @svnlooklines = &read_from_process($svnlook, 'changed', $repos, '-r', $rev);
 
 # Parse the changed nodes.
+my @fileschanged;
 my @adds;
 my @dels;
 my @mods;
+my $rootchanged = 0;
 foreach my $line (@svnlooklines)
   {
     my $path = '';
@@ -261,6 +245,12 @@
         $path = $2;
       }
 
+    push(@fileschanged, $path);
+    if ($path eq '/')
+      {
+        $rootchanged = 1;
+      }
+
     if ($code eq 'A')
       {
         push(@adds, $path);
@@ -288,11 +278,11 @@
 # there's no point in collapsing the directories, and only if more
 # than one directory was modified.
 my $commondir = '';
-if (!$rootchanged and @dirschanged > 1)
+if (!$rootchanged and @fileschanged > 1)
   {
-    my $firstline    = shift @dirschanged;
+    my $firstline    = shift @fileschanged;
     my @commonpieces = split('/', $firstline);
-    foreach my $line (@dirschanged)
+    foreach my $line (@fileschanged)
       {
         my @pieces = split('/', $line);
         my $i = 0;
@@ -306,28 +296,28 @@
             $i++;
           }
       }
-    unshift(@dirschanged, $firstline);
+    unshift(@fileschanged, $firstline);
 
     if (@commonpieces)
       {
         $commondir = join('/', @commonpieces);
-        my @new_dirschanged;
-        foreach my $dir (@dirschanged)
+        my @new_fileschanged;
+        foreach my $file (@fileschanged)
           {
-            if ($dir eq $commondir)
+            if ($file eq "$commondir/")
               {
-                $dir = '.';
+                $file = '.';
               }
             else
               {
-                $dir =~ s#^$commondir/##;
+                $file =~ s#^$commondir/##;
               }
-            push(@new_dirschanged, $dir);
+            push(@new_fileschanged, $file);
           }
-        @dirschanged = @new_dirschanged;
+        @fileschanged = @new_fileschanged;
       }
   }
-my $dirlist = join(' ', @dirschanged);
+my $filelist = join(' ', @fileschanged);
 
 ######################################################################
 # Assembly of log message.
@@ -367,7 +357,7 @@
   {
     my $match_re = $project->{match_re};
     my $match    = 0;
-    foreach my $path (@dirschanged, @adds, @dels, @mods)
+    foreach my $path (@adds, @dels, @mods)
       {
         if ($path =~ $match_re)
           {
@@ -390,16 +380,20 @@
 
     if ($commondir ne '')
       {
-        $subject = "r$rev - in $commondir: $dirlist";
+        $subject = "r$rev - in $commondir: $filelist";
       }
     else
       {
-        $subject = "r$rev - $dirlist";
+        $subject = "r$rev - $filelist";
       }
     if ($subject_prefix =~ /\w/)
       {
         $subject = "$subject_prefix $subject";
       }
+    if(length($subject) > 900)
+      {
+        $subject = substr($subject, 0, 896) . " ...";
+      }
     my $mail_from = $author;
 
     if ($from_address =~ /\w/)
